modules/internal/ChapelArray.chpl

```diff
@@ -3211,6 +3211,68 @@ module ChapelArray {
       else
         a.add(e);
   }
+
+  //
+  // BaseSparseDom operator overloads
+  //
+  proc +=(ref sd: domain, inds: [] sd.idxType) where isSparseDom(sd) &&
+    sd.rank==1 {
+
+    if inds.size == 0 then return;
+
+    sd._value.dsiBulkAdd(inds);
+  }
+
+  proc +=(ref sd: domain, inds: [] sd.rank*sd.idxType) where isSparseDom(sd) &&
+    sd.rank>1 {
+
+    if inds.size == 0 then return;
+
+    sd._value.dsiBulkAdd(inds);
+  }
+
+  // Currently this is not optimized for addition of a sparse
+  // This assumes that a domain is sorted by row, which may not always be true.
+  // There could be domains sorted by column instead.
+  // It would be nice if there was a common interface in BaseDom to query
+  // whether a domain was row-major or column-major to distinguish these.
+  // The current implementation is a work-around for this shortcoming.
+  proc +=(ref sd: domain, d: domain)
+  where sd._value: DefaultSparseDom && d.rank==sd.rank && sd.idxType==d.idxType {
+
+    if d.size == 0 then return;
+
+    type _idxType = if sd.rank==1 then int else sd.rank*int;
+    const indCount = d.numIndices;
+    const arr: [{0..#indCount}] _idxType;
+
+    //this could be a parallel loop. but ranks don't match -- doesn't compile
+    for (a,i) in zip(arr,d) do a=i;
```

Comment:
Even if ranks matched, this would be difficult to execute in parallel, since it would require us to be able to know a priori what the unique position of any index 'i' in domain 'd' had (i.e., what index of 'arr' each index in 'd' corresponds with) which is a hard problem in general.  I.e., I'm not sure "could be a parallel loop" is accurate without some sort of pre-computation of how many indices are in any given range of rows.

I felt surprised that we permit rank mismatches in serial loops like this, and am thinking that this might be one of the issues leader-follower 2.0 is supposed to fix.  Since it's already serial, it might be better to zip `d` with `0..` in order to make it future-proof.  (i.e., I'm a little worried that what you've written might break in the future).  That is:

```chapel
for (i,j) in zip(d, 0..) do arr[j] = i;
```